### ISSUE TITLE
Patches needed for resuming training

### ISSUE THREAD

**Issue by kanishkg** [willccbb/verifiers#129]
The current trainer and data loader doesn't expose a couple of attributes and don't handle an edge case which causes the training to fail. A couple of small patches fix this (I can create a pull request if this is useful to others):

```
# verifiers/trainers/grpo_trainer.py
def _gather_batch_data(self, batch_offset: int = 0):
    """
    Gather batch data from all processes.

    Args:
        batch_offset: 0 for the next batch, >0 to peek further ahead
    """
    # -------- FIX: make sure we actually fetch a batch when offset == 0 -----
    peek_n  = 1 if batch_offset == 0 else batch_offset
    batches = self._async_dataloader.peek_ahead(peek_n)

    # If nothing is available (e.g. dataloader exhausted) just return empties
    if not batches:
        return [], [], [], []

    batch = batches[batch_offset - 1] if batch_offset > 0 else batches[0]
    # ------------------------------------------------------------------------
```

and 

```
class AsyncDataLoaderWrapper:

    @property
    def sampler(self):
        return self.dataloader.sampler          # Needed by skip_first_batches

    @property
    def batch_sampler(self):
        return self.dataloader.batch_sampler    # Also accessed later

    # or this more general one that can be future proof
    def __getattr__(self, name):
        """
        Fallback: delegate any other attribute look-ups to the wrapped
        DataLoader.
        """
        return getattr(self.dataloader, name)

**Comment by willccbb** [comment]
do you know if this still an issue? did some reworking of the trainer to address the edge cases I was aware of, but haven't tested resuming from checkpoints.

I will note that the way we're doing async training isn't the most conducive to resuming from checkpoints--after step 1 of training, it's always assumed that inference is being done in parallel (if max_steps_async >0), and so resuming has to deal with the fact that there are 2 parallel datastreams -- prompts to the inference worker and completions to the trainer -- which need not be synchromized. My practical recommendation would be to train with a constant LR (maybe a short warmup), and to reshuffle your data (GRPO is pretty resilient to prompt reuse without overfitting). Robust support for checkpoint resuming is low on my current feature priority list, but happy to consider a PR if there's a clean way to fix it without adding too much complexity.
